### ISSUE TITLE
Task vectors for GPT2 model: attn.bias weights ignored

### ISSUE THREAD

**Issue by adibMosharrof** [mlfoundations/task_vectors#2]
I am trying to use task vectors for GPT2 models.

In the TaskVectors class, when the task vector is created, there is a [condition](https://github.com/mlfoundations/task_vectors/blob/main/src/task_vectors.py#L23) which ignores keys in the state dict that have dtype of uint8.

Due to this condition, when I call the apply_to() method of a task vector instance by passing a GPT-2 model checkpoint, I get the following error.

`Warning: key transformer.h.0.attn.bias is present in the pretrained state dict but not in the task vector
Warning: key transformer.h.1.attn.bias is present in the pretrained state dict but not in the task vector
Warning: key transformer.h.2.attn.bias is present in the pretrained state dict but not in the task vector
Warning: key transformer.h.3.attn.bias is present in the pretrained state dict but not in the task vector
Warning: key transformer.h.4.attn.bias is present in the pretrained state dict but not in the task vector
Warning: key transformer.h.5.attn.bias is present in the pretrained state dict but not in the task vector
Warning: key transformer.h.6.attn.bias is present in the pretrained state dict but not in the task vector
Warning: key transformer.h.7.attn.bias is present in the pretrained state dict but not in the task vector
Warning: key transformer.h.8.attn.bias is present in the pretrained state dict but not in the task vector
Warning: key transformer.h.9.attn.bias is present in the pretrained state dict but not in the task vector
Warning: key transformer.h.10.attn.bias is present in the pretrained state dict but not in the task vector
Warning: key transformer.h.11.attn.bias is present in the pretrained state dict but not in the task vector`

Is there a reason why state dict keys with dtype torch.uint8  are ignored. 

When that condition is removed, the code to run without any errors. 

Please suggest what should be the best thing to do here.

**Comment by gabrielilharco** [comment]
Hi @adibMosharrof are the biases in your checkpoint `uint8`s?

I'm seeing all parameters as float32 when loading a checkpoint from HF for instance, e.g. here is the output of this code snippet:

```
>>> from transformers import GPT2Model                                                                                                                                                                     
>>> model = GPT2Model.from_pretrained("gpt2")                                                                                                                                                              
>>> for name, param in model.named_parameters(): print(param.dtype, name) 
```               

```
torch.float32 wte.weight                                                                                                                                                                                   
torch.float32 wpe.weight                                                                                                                                                                                   
torch.float32 h.0.ln_1.weight                                                                                                                                                                              
torch.float32 h.0.ln_1.bias                                                                                                                                                                                
torch.float32 h.0.attn.c_attn.weight                                                                                                                                                                       
torch.float32 h.0.attn.c_attn.bias                                                                                                                                                                         
torch.float32 h.0.attn.c_proj.weight                                                                                                                                                                       
torch.float32 h.0.attn.c_proj.bias                                                                                                                                                                         
torch.float32 h.0.ln_2.weight                                                                                                                                                                              
torch.float32 h.0.ln_2.bias                                                                                                                                                                                
torch.float32 h.0.mlp.c_fc.weight                                                                                                                                                                          
torch.float32 h.0.mlp.c_fc.bias                                                                                                                                                                            
torch.float32 h.0.mlp.c_proj.weight                                                                                                                                                                        
torch.float32 h.0.mlp.c_proj.bias                                                                                                                                                                          
torch.float32 h.1.ln_1.weight                                                                                                                                                                              
torch.float32 h.1.ln_1.bias                                                                                                                                                                                
torch.float32 h.1.attn.c_attn.weight                                                                                                                                                                       
torch.float32 h.1.attn.c_attn.bias                                                                                                                                                                         
torch.float32 h.1.attn.c_proj.weight                                                                                                                                                                       
torch.float32 h.1.attn.c_proj.bias
torch.float32 h.1.ln_2.weight
torch.float32 h.1.ln_2.bias
torch.float32 h.1.mlp.c_fc.weight
torch.float32 h.1.mlp.c_fc.bias
torch.float32 h.1.mlp.c_proj.weight
torch.float32 h.1.mlp.c_proj.bias
torch.float32 h.2.ln_1.weight
torch.float32 h.2.ln_1.bias
torch.float32 h.2.attn.c_attn.weight
torch.float32 h.2.attn.c_attn.bias
torch.float32 h.2.attn.c_proj.weight
torch.float32 h.2.attn.c_proj.bias
torch.float32 h.2.ln_2.weight
torch.float32 h.2.ln_2.bias
torch.float32 h.2.mlp.c_fc.weight
torch.float32 h.2.mlp.c_fc.bias
torch.float32 h.2.mlp.c_proj.weight
torch.float32 h.2.mlp.c_proj.bias
...
torch.float32 h.10.ln_1.weight
torch.float32 h.10.ln_1.bias
torch.float32 h.11.attn.c_attn.weight
torch.float32 h.11.attn.c_attn.bias
torch.float32 h.11.attn.c_proj.weight
torch.float32 h.11.attn.c_proj.bias
torch.float32 h.11.ln_2.weight
torch.float32 h.11.ln_2.bias
torch.float32 h.11.mlp.c_fc.weight
torch.float32 h.11.mlp.c_fc.bias
torch.float32 h.11.mlp.c_proj.weight
torch.float32 h.11.mlp.c_proj.bias
torch.float32 ln_f.weight
torch.float32 ln_f.bias
```                                                                                                             
...

**Comment by gabrielilharco** [comment]
After looking a bit more into it, it seems like things like `transformer.h.0.attn.bias` are buffers for the causal mask, and are not modified during training (see https://github.com/huggingface/transformers/blob/v4.26.1/src/transformers/models/gpt2/modeling_gpt2.py#L127).

It's fine to ignore those parameters, they should be equal for any GPT2 checkpoint

**Comment by adibMosharrof** [comment]
@gabrielilharco Thank you very much for such a quick response. 
You are right, the h attn bias is uint8. 
To prevent the code from failing, I have the following lines commented out from the task_vectors.py file.
`
if pretrained_state_dict[key].dtype == torch.uint8:
    continue
`

Should I keep these lines commented out, or should I manually enter those biases from the code link you provided to the GPT2 model inside the apply_to() function in task_vectors.py.

**Comment by adibMosharrof** [comment]
Another think to point out is that, in your comment you used 
`model.named_parameters()`
whereas in the task_vectors.py, apply_to() function, the parameters are obtained using
`pretrained_model.state_dict()`
I really don't know if there are major differences in these functions.

**Comment by gabrielilharco** [comment]
Commenting that if statement should be fine. To be sure, can you check that `finetuned_state_dict[key]` and `pretrained_state_dict[key]` are the same for those cases?
This would mean that applying a task vector won't alter those parameters, which is what we desire.

`named_parameters()` and `state_dict()` are indeed different (see [here](https://stackoverflow.com/questions/54746829/pytorch-whats-the-difference-between-state-dict-and-parameters)), `state_dict` also contains things like buffers (which is what those `attn.bias` tensors are).

Also, to clarify, are you getting errors? Or just warnings?

**Comment by adibMosharrof** [comment]
I will take a look at the things you suggested and get back to you.

I pasted the warnings from the program output, but I get errors with those messages on this line
`pretrained_model.load_state_dict(new_state_dict)`

**Comment by gabrielilharco** [comment]
Got it, you should also be able to bypass the error by using `strict=False` when calling `load_state_dict`

**Comment by adibMosharrof** [comment]
Using `strict=False` worked. Thank you very much for your support :+1:
